File: testing/tests_js/test_023_deferred_operations.py
"""Node-backed checks for deferred-operation poll subscriptions."""


# @matrix deferred-jobs : backoff decoration-opt-out lazy-watcher polling progress rendered-visibility revision status teardown terminal-ownership timing visible-blur
def test_deferred_operation_manager_batches_orders_and_renders_status(run_node):
    run_node(
        r"""
const fs = require("node:fs");
const vm = require("node:vm");

function operationNode(key, visible = true) {
	const phase = { textContent: "Waiting to start" };
	const elapsed = { textContent: "just now" };
	return {
		dataset: { operation: key, operationRevision: "0" },
		isConnected: true,
		visible,
		phase,
		elapsed,
		closest() { return this.visible ? null : {}; },
		getClientRects() { return this.visible ? [{}] : []; },
    querySelector(selector) {
      if (selector === "[data-role='deferred-phase']") return phase;
      if (selector === "[data-role='deferred-elapsed']") return elapsed;
      return null;
    },
    querySelectorAll() { return []; },
  };
}

const nodes = [operationNode("operation-a"), operationNode("operation-b", false)];
const subscriptions = new Map();
const descriptors = new Map();
const schedules = new Map();
const triggers = [];
const reconciled = [];
const expectedCompletions = [];
let ensureEditWatcherCalls = 0;
let reschedules = 0;
const blurStarts = [];
const coordinator = {
  subscribe(descriptor, hooks) {
    descriptors.set(descriptor.id, descriptor);
    schedules.set(descriptor.id, {
      mode: hooks.mode,
      initial: hooks.initial,
    });
    subscriptions.set(descriptor.id, hooks);
    return () => subscriptions.delete(descriptor.id);
  },
	async trigger(ids) { triggers.push(ids); return []; },
	blur(startedAt) { blurStarts.push(startedAt); },
	reschedule() { reschedules += 1; },
};
const context = {
  console,
  CustomEvent: class {
    constructor(type, options) { this.type = type; this.detail = options.detail; }
  },
  createIcon() { return { setAttribute() {} }; },
  withTransition(callback) { return callback(); },
  document: {
    createElement() {
      return { dataset: {}, children: [], setAttribute() {},
        append(...children) { this.children.push(...children); } };
    },
    querySelectorAll(selector) {
      return selector === "[data-operation]" ? nodes : [];
    },
  },
  window: { dispatchEvent() {} },
};
vm.createContext(context);
let source = fs.readFileSync("src/script/shared/deferredOperations.mjs", "utf8");
source = source.replace(/^import .*;\n/gm, "");
source = source.replace(
  "export class DeferredOperationManager",
  "class DeferredOperationManager",
);
source += "\nglobalThis.DeferredOperationManager = DeferredOperationManager;";
vm.runInContext(source, context);

const editWatcher = {
  expectDeferredCompletion(key, operation) {
    expectedCompletions.push({ key, operation });
  },
};
const view = {
  PollingCoordinator: coordinator,
  EditWatcher: null,
  async ensureEditWatcher() {
    ensureEditWatcherCalls += 1;
    this.EditWatcher = editWatcher;
    return editWatcher;
  },
  async reconcileChange(change) { reconciled.push(change); },
};
const manager = new context.DeferredOperationManager(view).init();

(async () => {
  if (subscriptions.size !== 2) {
    throw new Error("Visible operations did not create polling subscriptions");
  }
	if (triggers.length !== 0 ||
			descriptors.get("operation:operation-a")?.revision !== 0 ||
			schedules.get("operation:operation-a")?.initial !== "scheduled") {
		throw new Error("Server-rendered operation did not seed a delayed revision cursor");
	}
	if (
		subscriptions.get("operation:operation-a").whileBlurred() !== true ||
		subscriptions.get("operation:operation-b").whileBlurred() !== false
	) {
		throw new Error("Blurred polling did not follow rendered operation visibility");
	}
	manager.track("operation-local", { revision: 2 });
	if (triggers.length !== 1 || triggers[0] !== "operation:operation-local") {
		throw new Error("Locally started operation did not request an immediate poll");
	}
	if (subscriptions.get("operation:operation-local").whileBlurred() !== false) {
		throw new Error("A background-only operation qualified for visible blur polling");
	}
	view.hidden = true;
	view.blurred = true;
	view.blurredAt = 42;
	manager.track("operation-a", { node: nodes[0], immediate: false });
	view.hidden = false;
	view.blurred = false;
	if (reschedules !== 1 || blurStarts.join(",") !== "42") {
		throw new Error("Existing operation visibility did not refresh scheduling");
	}
  await manager.poll();
  if (triggers.length !== 2 || triggers[1].length !== 3) {
    throw new Error("Operation poll did not delegate a batch to the coordinator");
  }

  await subscriptions.get("operation:operation-a").onResult({
    status: "changed",
    payload: {
      key: "operation-a",
      status: "running",
      phase: "generating",
      phase_label: "Generating",
      elapsed_seconds: 75,
      revision: 3,
      terminal: false,
    },
  });
  if (nodes[0].phase.textContent !== "Generating" ||
      nodes[0].elapsed.textContent !== "1 min") {
    throw new Error("Active operation status was not rendered");
  }

  await subscriptions.get("operation:operation-b").onResult({
    status: "changed",
    payload: {
      key: "operation-b",
      status: "succeeded",
      phase: "complete",
      phase_label: "Complete",
      elapsed_seconds: 12,
      revision: 2,
      terminal: true,
      entity_key: "report-b",
      source_widget: "CreateToolReport",
      destination: "tools:ToolReportList",
    },
  });
  if (manager.operations.has("operation-b") ||
      reconciled[0]?.key !== "report-b") {
    throw new Error("Terminal operation was not reconciled and retired");
  }
  if (
    ensureEditWatcherCalls !== 1 ||
    expectedCompletions[0]?.key !== "report-b" ||
    expectedCompletions[0]?.operation !== "operation-b"
  ) {
    throw new Error("Terminal operation did not await ownership reconciliation");
  }
  if (manager.nudge("operation-a", 2) !== false) {
    throw new Error("An out-of-order operation revision was accepted");
  }

  const newer = operationNode("operation-a");
  Object.assign(newer.dataset, {
    operationRevision: "5", operationStatus: "retry_wait",
    operationPhase: "using_tools", operationPhaseLabel: "Checking context",
    operationRecovering: "true", operationElapsed: "80",
  });
  const older = operationNode("operation-a");
  Object.assign(older.dataset, {
    operationRevision: "2", operationStatus: "running",
    operationPhase: "generating", operationPhaseLabel: "Generating",
  });
  nodes.push(newer, older);
  manager.scan();
  for (const node of [nodes[0], newer, older]) {
    if (node.phase.textContent !== "Checking context. Automatic recovery is active." ||
        node.dataset.operationRevision !== "5") {
      throw new Error("An older sibling snapshot replaced the shared latest status");
    }
  }
  manager.scan();
  await subscriptions.get("operation:operation-a").onResult({ status: "unchanged" });
  if (older.phase.textContent !== newer.phase.textContent ||
      manager.operations.get("operation-a").status.phase_label !== "Checking context") {
    throw new Error("Rescanning rendered progress lost the cached phase metadata");
  }
  manager.nudge("operation-a");
  if (manager.operations.get("operation-a").revision !== 5) {
    throw new Error("A cursorless nudge erased the shared operation revision");
  }
  manager.track("operation-replacement", { node: newer, immediate: false });
  if (!subscriptions.has("operation:operation-a") || older.dataset.operation !== "operation-a") {
    throw new Error("Reusing one target retired a job still owned by another target");
  }

  const owner = operationNode("operation-owned");
  owner.querySelector = (selector) => owner.progress?.children.find(
    (child) => selector === "[data-role='deferred-phase']" && child.dataset?.role === "deferred-phase",
  );
  owner.append = (progress) => { owner.progress = progress; };
  nodes.push(owner);
  manager.track("operation-owned", { node: owner, immediate: false });
  if (!owner.progress || owner.progress.dataset.operation) {
    throw new Error("Generated progress became a second operation owner");
  }
  manager.track("operation-next", { node: owner, immediate: false });
  if (subscriptions.has("operation:operation-owned") || owner.dataset.operation !== "operation-next") {
    throw new Error("Generated progress kept a replaced operation subscribed");
  }

  manager.destroy();
  if (manager.operations.size || subscriptions.size) {
    throw new Error("Destroy did not clear operation state");
  }
})().catch((error) => {
  console.error(error);
  process.exit(1);
});
"""
    )


# @matrix deferred-jobs : terminal-ownership revision
def test_deferred_operation_manager_reconciles_server_rendered_terminal_status(
    run_node,
):
    run_node(
        r"""
const fs = require("node:fs");
const vm = require("node:vm");

const phase = { textContent: "Complete" };
const elapsed = { textContent: "just now" };
const node = {
  dataset: {
    key: "report-ready",
    operation: "operation-complete",
    operationRevision: "7",
    operationStatus: "succeeded",
    operationPhase: "complete",
    operationPhaseLabel: "Complete",
    operationElapsed: "3",
    operationRecovering: "false",
    operationTerminal: "true",
  },
  isConnected: true,
  closest() { return null; },
  getClientRects() { return [{}]; },
  querySelector(selector) {
    if (selector === "[data-role='deferred-phase']") return phase;
    if (selector === "[data-role='deferred-elapsed']") return elapsed;
    return null;
  },
  querySelectorAll() { return []; },
};
const subscriptions = new Map();
const triggers = [];
const events = [];
const reconciled = [];
const expectedCompletions = [];
const context = {
  console,
  setTimeout,
  CustomEvent: class {
    constructor(type, options) { this.type = type; this.detail = options.detail; }
  },
  createIcon() { return {}; },
  withTransition(callback) { return callback(); },
  document: {
    querySelectorAll(selector) {
      return selector === "[data-operation]" ? [node] : [];
    },
  },
  window: { dispatchEvent(event) { events.push(event); } },
};
vm.createContext(context);
let source = fs.readFileSync("src/script/shared/deferredOperations.mjs", "utf8");
source = source.replace(/^import .*;\n/gm, "");
source = source.replace(
  "export class DeferredOperationManager",
  "class DeferredOperationManager",
);
source += "\nglobalThis.DeferredOperationManager = DeferredOperationManager;";
vm.runInContext(source, context);

const view = {
  PollingCoordinator: {
    subscribe(descriptor, hooks) {
      subscriptions.set(descriptor.id, { descriptor, ...hooks });
      return () => subscriptions.delete(descriptor.id);
    },
    trigger(ids) { triggers.push(ids); },
    reschedule() {},
  },
  EditWatcher: {
    expectDeferredCompletion(key, operation) {
      expectedCompletions.push({ key, operation });
    },
  },
  async reconcileChange(change) {
    reconciled.push(change);
    if (reconciled.length === 1) throw new Error("Replacement temporarily unavailable");
  },
};

(async () => {
  const manager = new context.DeferredOperationManager(view).init();
  manager.scan();
  const subscription = subscriptions.get("operation:operation-complete");
  if (events.length || reconciled.length || subscriptions.size !== 1 ||
      subscription.descriptor.revision >= 7 || triggers.length !== 2) {
    throw new Error("Terminal HTML bypassed the authoritative operation poll");
  }
  const result = {
    status: "changed",
    payload: {
      key: "operation-complete", revision: 7, status: "succeeded",
      phase: "complete", phase_label: "Complete", terminal: true,
      entity_key: "authoritative-report", source_widget: "CreateToolReport",
      destination: "tools:ToolReportList",
    },
  };
  if (await subscription.onResult(result) !== false || subscriptions.size !== 1) {
    throw new Error("Failed terminal replacement was acknowledged instead of retried");
  }
  if (await subscription.onResult(result) !== true) {
    throw new Error("Terminal replacement could not recover at the same revision");
  }

  if (
    events.length !== 2 ||
    events[0].type !== "deferred-operation" ||
    events[0].detail.key !== "operation-complete" ||
    events[0].detail.entity_key !== "authoritative-report"
  ) {
    throw new Error("Server-rendered terminal status was not published to its view owner");
  }
  if (
    reconciled.length !== 2 ||
    reconciled[0].type !== "deferred-complete" ||
    reconciled[0].key !== "authoritative-report" ||
    reconciled[0].destination !== "tools:ToolReportList" ||
    reconciled[0].source_widget !== "CreateToolReport" ||
    expectedCompletions[0]?.operation !== "operation-complete"
  ) {
    throw new Error("Server-rendered terminal status was not reconciled");
  }
  if (manager.operations.size || subscriptions.size) {
    throw new Error("Reconciled terminal seed remained subscribed");
  }
})().catch((error) => {
  console.error(error);
  process.exit(1);
});
"""
    )

File: testing/tests_js/test_023_deferred_operations.mjs
import assert from "node:assert/strict";
import { test } from "node:test";
import esmock from "esmock";
import { createBrowser } from "../utility/js/environment.mjs";

function operationNode(key, { visible = true, append = true } = {}) {
	const node = document.createElement("section");
	node.dataset.operation = key;
	node.dataset.operationRevision = "0";
	const phase = document.createElement("span");
	phase.dataset.role = "deferred-phase";
	phase.textContent = "Waiting to start";
	const elapsed = document.createElement("span");
	elapsed.dataset.role = "deferred-elapsed";
	elapsed.textContent = "just now";
	node.append(phase, elapsed);
	node.closest = () => (visible ? null : document.body);
	node.getClientRects = () => (visible ? [{}] : []);
	if (append) document.body.append(node);
	return { elapsed, node, phase };
}

async function loadManager() {
	const capturedErrors = [];
	const createIcon = () => document.createElement("span");
	const withTransition = (callback) => callback();
	const { DeferredOperationManager } = await esmock.strict(
		"../../src/script/shared/deferredOperations.mjs",
		{
			"../../src/script/shared/errors.mjs": {
				captureError: (...args) => capturedErrors.push(args),
			},
			"../../src/script/shared/icons.mjs": { createIcon },
			"../../src/script/shared/transitions.mjs": { withTransition },
		},
	);
	return { capturedErrors, DeferredOperationManager };
}

/**
 * @matrix deferred-jobs : backoff decoration-opt-out lazy-watcher polling progress rendered-visibility revision status teardown terminal-ownership timing visible-blur
 */
test("test_deferred_operation_manager_batches_orders_and_renders_status", async (t) => {
	createBrowser(t);
	const first = operationNode("operation-a");
	operationNode("operation-b", { visible: false });
	const subscriptions = new Map();
	const descriptors = new Map();
	const schedules = new Map();
	const triggers = [];
	const reconciled = [];
	const expectedCompletions = [];
	let ensureEditWatcherCalls = 0;
	let reschedules = 0;
	const blurStarts = [];
	const coordinator = {
		subscribe(descriptor, hooks) {
			descriptors.set(descriptor.id, descriptor);
			schedules.set(descriptor.id, {
				mode: hooks.mode,
				initial: hooks.initial,
			});
			subscriptions.set(descriptor.id, hooks);
			return () => subscriptions.delete(descriptor.id);
		},
		async trigger(ids) {
			triggers.push(ids);
			return [];
		},
		blur(startedAt) {
			blurStarts.push(startedAt);
		},
		reschedule() {
			reschedules += 1;
		},
	};
	const { DeferredOperationManager } = await loadManager();
	const editWatcher = {
		expectDeferredCompletion(key, operation) {
			expectedCompletions.push({ key, operation });
		},
	};
	const view = {
		PollingCoordinator: coordinator,
		EditWatcher: null,
		async ensureEditWatcher() {
			ensureEditWatcherCalls += 1;
			this.EditWatcher = editWatcher;
			return editWatcher;
		},
		async reconcileChange(change) {
			reconciled.push(change);
		},
	};
	const manager = new DeferredOperationManager(view).init();

	assert.equal(subscriptions.size, 2);
	assert.equal(triggers.length, 0);
	assert.equal(descriptors.get("operation:operation-a")?.revision, 0);
	assert.equal(schedules.get("operation:operation-a")?.initial, "scheduled");
	assert.equal(subscriptions.get("operation:operation-a").whileBlurred(), true);
	assert.equal(
		subscriptions.get("operation:operation-b").whileBlurred(),
		false,
	);
	manager.track("operation-local", { revision: 2 });
	assert.equal(triggers[0], "operation:operation-local");
	assert.equal(
		subscriptions.get("operation:operation-local").whileBlurred(),
		false,
	);
	view.hidden = true;
	view.blurred = true;
	view.blurredAt = 42;
	manager.track("operation-a", { node: first.node, immediate: false });
	view.hidden = false;
	view.blurred = false;
	assert.equal(reschedules, 1);
	assert.deepEqual(blurStarts, [42]);
	await manager.poll();
	assert.equal(triggers.length, 2);
	assert.equal(triggers[1].length, 3);

	await subscriptions.get("operation:operation-a").onResult({
		status: "changed",
		payload: {
			key: "operation-a",
			status: "running",
			phase: "generating",
			phase_label: "Generating",
			elapsed_seconds: 75,
			revision: 3,
			terminal: false,
		},
	});
	assert.equal(first.phase.textContent, "Generating");
	assert.equal(first.elapsed.textContent, "1 min");

	await subscriptions.get("operation:operation-b").onResult({
		status: "changed",
		payload: {
			key: "operation-b",
			status: "succeeded",
			phase: "complete",
			phase_label: "Complete",
			elapsed_seconds: 12,
			revision: 2,
			terminal: true,
			entity_key: "report-b",
			source_widget: "CreateToolReport",
			destination: "tools:ToolReportList",
		},
	});
	assert.equal(manager.operations.has("operation-b"), false);
	assert.equal(reconciled[0]?.key, "report-b");
	assert.equal(ensureEditWatcherCalls, 1);
	assert.deepEqual(expectedCompletions[0], {
		key: "report-b",
		operation: "operation-b",
	});
	assert.equal(manager.nudge("operation-a", 2), false);

	const newer = operationNode("operation-a");
	Object.assign(newer.node.dataset, {
		operationRevision: "5",
		operationStatus: "retry_wait",
		operationPhase: "using_tools",
		operationPhaseLabel: "Checking context",
		operationRecovering: "true",
		operationElapsed: "80",
	});
	const older = operationNode("operation-a");
	Object.assign(older.node.dataset, {
		operationRevision: "2",
		operationStatus: "running",
		operationPhase: "generating",
		operationPhaseLabel: "Generating",
	});
	manager.scan();
	for (const operation of [first, newer, older]) {
		assert.equal(
			operation.phase.textContent,
			"Checking context. Taking longer than expected.",
		);
		assert.equal(operation.node.dataset.operationRevision, "5");
	}
	manager.scan();
	await subscriptions
		.get("operation:operation-a")
		.onResult({ status: "unchanged" });
	assert.equal(older.phase.textContent, newer.phase.textContent);
	assert.equal(
		manager.operations.get("operation-a").status.phase_label,
		"Checking context",
	);
	manager.nudge("operation-a");
	assert.equal(manager.operations.get("operation-a").revision, 5);
	manager.track("operation-replacement", {
		node: newer.node,
		immediate: false,
	});
	assert.equal(subscriptions.has("operation:operation-a"), true);
	assert.equal(older.node.dataset.operation, "operation-a");

	const owner = document.createElement("section");
	owner.dataset.operation = "operation-owned";
	owner.dataset.operationRevision = "0";
	owner.closest = () => null;
	owner.getClientRects = () => [{}];
	document.body.append(owner);
	manager.track("operation-owned", { node: owner, immediate: false });
	const progress = owner.querySelector("[data-role='deferred-progress']");
	assert.ok(progress);
	assert.equal(Boolean(progress.dataset.operation), false);
	manager.track("operation-next", { node: owner, immediate: false });
	assert.equal(subscriptions.has("operation:operation-owned"), false);
	assert.equal(owner.dataset.operation, "operation-next");

	for (const [scope, expected] of [
		["form-change", "Schema migration in progress"],
		["form-autofill", "Autofill queued"],
	]) {
		const form = document.createElement("form");
		form.dataset.operation = `operation-${scope}`;
		form.dataset.operationRevision = "0";
		form.dataset.deferredLock = "form";
		form.dataset.operationScope = scope;
		const autofill = document.createElement("button");
		autofill.dataset.role = "autofill";
		const submit = document.createElement("div");
		submit.dataset.role = "submit-group";
		form.append(autofill, submit);
		document.body.append(form);
		manager.scan({ querySelectorAll: () => [form] });
		assert.equal(
			form.querySelector("[data-role='deferred-phase']")?.textContent,
			expected,
		);
		await manager.receive({
			key: form.dataset.operation,
			revision: 1,
			status: "running",
			phase_label: "Applying changes",
			terminal: false,
		});
		assert.equal(
			form.querySelector("[data-role='deferred-phase']").textContent,
			"Applying changes",
		);
	}

	manager.destroy();
	assert.equal(manager.operations.size, 0);
	assert.equal(subscriptions.size, 0);
});

/** @matrix deferred-jobs : terminal-ownership revision */
test("test_deferred_operation_manager_reconciles_server_rendered_terminal_status", async (t) => {
	createBrowser(t);
	const operation = operationNode("operation-complete");
	Object.assign(operation.node.dataset, {
		key: "report-ready",
		operationRevision: "7",
		operationStatus: "succeeded",
		operationPhase: "complete",
		operationPhaseLabel: "Complete",
		operationElapsed: "3",
		operationRecovering: "false",
		operationTerminal: "true",
	});
	operation.phase.textContent = "Complete";
	const subscriptions = new Map();
	const triggers = [];
	const events = [];
	const reconciled = [];
	const expectedCompletions = [];
	window.addEventListener("deferred-operation", (event) => events.push(event));
	const { DeferredOperationManager } = await loadManager();
	const view = {
		PollingCoordinator: {
			subscribe(descriptor, hooks) {
				subscriptions.set(descriptor.id, { descriptor, ...hooks });
				return () => subscriptions.delete(descriptor.id);
			},
			trigger(ids) {
				triggers.push(ids);
			},
			reschedule() {},
		},
		EditWatcher: {
			expectDeferredCompletion(key, operationKey) {
				expectedCompletions.push({ key, operation: operationKey });
			},
		},
		async reconcileChange(change) {
			reconciled.push(change);
			if (reconciled.length === 1) {
				throw new Error("Replacement temporarily unavailable");
			}
		},
	};
	const manager = new DeferredOperationManager(view).init();
	manager.scan();
	const subscription = subscriptions.get("operation:operation-complete");
	assert.equal(events.length, 0);
	assert.equal(reconciled.length, 0);
	assert.equal(subscriptions.size, 1);
	assert.ok(subscription.descriptor.revision < 7);
	assert.equal(triggers.length, 2);
	const result = {
		status: "changed",
		payload: {
			key: "operation-complete",
			revision: 7,
			status: "succeeded",
			phase: "complete",
			phase_label: "Complete",
			terminal: true,
			entity_key: "authoritative-report",
			source_widget: "CreateToolReport",
			destination: "tools:ToolReportList",
		},
	};
	assert.equal(await subscription.onResult(result), false);
	assert.equal(subscriptions.size, 1);
	assert.equal(await subscription.onResult(result), true);
	assert.equal(events.length, 2);
	assert.equal(events[0].type, "deferred-operation");
	assert.equal(events[0].detail.key, "operation-complete");
	assert.equal(events[0].detail.entity_key, "authoritative-report");
	assert.equal(reconciled.length, 2);
	assert.equal(reconciled[0].type, "deferred-complete");
	assert.equal(reconciled[0].key, "authoritative-report");
	assert.equal(reconciled[0].destination, "tools:ToolReportList");
	assert.equal(reconciled[0].source_widget, "CreateToolReport");
	assert.equal(expectedCompletions[0]?.operation, "operation-complete");
	assert.equal(manager.operations.size, 0);
	assert.equal(subscriptions.size, 0);
});

/** @matrix deferred-jobs : rendered-autofill rendered-visibility */
test("test_cold_task_operations_wait_for_activation_and_completed_reviews_need_no_poll", async (t) => {
	createBrowser(t);
	const running = operationNode("task-running").node;
	running.dataset.widget = "TaskForm";
	running.dataset.operationRevision = "2";
	running.dataset.operationBootstrap = JSON.stringify({
		key: "task-running",
		revision: 2,
		type: "autofill",
		status: "running",
		terminal: false,
	});
	running.dataset.formState = JSON.stringify({
		operation: JSON.parse(running.dataset.operationBootstrap),
		reviews: [],
	});
	const complete = operationNode("task-complete").node;
	complete.dataset.widget = "TaskForm";
	complete.dataset.operationRevision = "7";
	complete.dataset.operationBootstrap = JSON.stringify({
		key: "task-complete",
		revision: 7,
		type: "autofill",
		status: "succeeded",
		terminal: true,
	});
	complete.dataset.formState = JSON.stringify({
		operation: JSON.parse(complete.dataset.operationBootstrap),
		reviews: [{ operation: "task-complete" }],
	});
	const subscriptions = new Map();
	const triggers = [];
	const { DeferredOperationManager } = await loadManager();
	const manager = new DeferredOperationManager({
		PollingCoordinator: {
			subscribe(descriptor, hooks) {
				subscriptions.set(descriptor.id, hooks);
				return () => subscriptions.delete(descriptor.id);
			},
			trigger(ids) {
				triggers.push(ids);
			},
			reschedule() {},
		},
	}).init();
	assert.equal(subscriptions.size, 0);
	assert.equal(triggers.length, 0);

	// Opening a task registers its running operation; closing drops it.
	manager.resumeTaskForm(running);
	assert.equal(subscriptions.has("operation:task-running"), true);
	assert.deepEqual(triggers, ["operation:task-running"]);
	manager.suspendTaskForm(running);
	assert.equal(subscriptions.size, 0);
	assert.equal(manager.operations.size, 0);
	manager.resumeTaskForm(running);
	assert.equal(subscriptions.has("operation:task-running"), true);

	// The server-rendered review is already authoritative; it needs no
	// operation poll or task-form reconciliation even after opening.
	manager.resumeTaskForm(complete);
	assert.equal(subscriptions.has("operation:task-complete"), false);
	assert.equal(triggers.includes("operation:task-complete"), false);
	manager.destroy();
});

/**
 * @source src/script/shared/deferredOperations.mjs::DeferredOperationManager
 * @pair deferred-jobs:review-probe
 */
test("test_successful_autofill_waits_for_active_form_review_before_retiring", async (t) => {
	createBrowser(t);
	const form = document.createElement("form");
	form.dataset.operation = "autofill-job";
	form.dataset.operationRevision = "0";
	document.body.append(form);
	const widget = { visible: true, reviewState: { reviews: [] } };
	widget.component = { active: widget };
	form._lp_widget = widget;
	const subscriptions = new Map();
	const { DeferredOperationManager } = await loadManager();
	const view = {
		PollingCoordinator: {
			subscribe(descriptor, hooks) {
				subscriptions.set(descriptor.id, hooks);
				return () => subscriptions.delete(descriptor.id);
			},
			reschedule() {},
		},
		EditWatcher: { expectDeferredCompletion() {} },
		async reconcileChange() {},
	};
	const manager = new DeferredOperationManager(view);
	manager.track("autofill-job", { node: form, immediate: false });
	const status = {
		key: "autofill-job",
		revision: 1,
		type: "autofill",
		status: "succeeded",
		terminal: true,
		entity_key: "task-key",
	};
	assert.equal(await manager.receive(status), false);
	assert.equal(manager.operations.has("autofill-job"), true);
	assert.equal(subscriptions.has("operation:autofill-job"), true);
	widget.reviewState.reviews = [{ operation: "autofill-job" }];
	assert.equal(await manager.receive(status), true);
	assert.equal(manager.operations.has("autofill-job"), false);
	assert.equal(subscriptions.has("operation:autofill-job"), false);
	form.dataset.operation = "obsolete-job";
	widget.reviewState.reviews = [];
	manager.track("obsolete-job", { node: form, immediate: false });
	const obsoleteStatus = { ...status, key: "obsolete-job" };
	assert.equal(await manager.receive(obsoleteStatus), false);
	widget.reviewState.stale_autofill = true;
	assert.equal(await manager.receive(obsoleteStatus), true);
	assert.equal(manager.operations.has("obsolete-job"), false);
	assert.equal(subscriptions.has("operation:obsolete-job"), false);
});
